mne/morph.py

```diff
@@ -1011,95 +1032,173 @@ def _compute_r2(a, b):
         (np.linalg.norm(a) * np.linalg.norm(b))
 
 
-def _compute_morph_sdr(mri_from, mri_to, niter_affine, niter_sdr, zooms):
+def _reslice_normalize(img, zooms):
+    from dipy.align.reslice import reslice
+    img_zooms = img.header.get_zooms()[:3]
+    img_affine = img.affine
+    img = _get_img_fdata(img)
+    if zooms is not None:
+        img, img_affine = reslice(img, img_affine, img_zooms, zooms)
+    img /= img.max()  # normalize
+    return img, img_affine
+
+
+_ORDERED_STEPS = ('center_of_mass', 'translation', 'rigid', 'affine', 'sdr')
+
+
+def _validate_niter(niter):
+    _validate_type(niter, (dict, list, tuple, None), 'niter')
+    niter = _handle_default('transform_niter', niter)
+    for key, value in niter.items():
+        _check_option('niter key', key, _ORDERED_STEPS[1:])
+        _check_option(f'len(niter[{repr(key)}])', len(value), (1, 2, 3))
+    return niter
+
+
+def _compute_morph_sdr(mri_from, mri_to, pipeline, niter, zooms,
+                       allow_separate_zooms=False):
     """Get a matrix that morphs data from one subject to another."""
+    _check_dep(nibabel='2.1.0', dipy='0.10.1')
+    from nibabel.spatialimages import SpatialImage
     with np.testing.suppress_warnings():
         from dipy.align import imaffine, imwarp, metrics, transforms
-    from dipy.align.reslice import reslice
 
-    logger.info('Computing nonlinear Symmetric Diffeomorphic Registration...')
+    #
+    # Input validation
+    #
+
+    _validate_type(mri_from, SpatialImage, 'moving image')
+    _validate_type(mri_to, SpatialImage, 'static image')
+
+    # niter, zooms
+    types = (list, tuple, 'numeric', None)
+    if allow_separate_zooms:
+        types = (dict,) + types
+    _validate_type(zooms, types, 'zooms')
+    zooms = _handle_default('transform_zooms', zooms)
+    for key, val in zooms.items():
+        _check_option('zooms key', key, ('affine', 'sdr'))
+        if val is not None:
+            val = tuple(
+                float(x) for x in np.array(val, dtype=float).ravel())
+            _check_option(f'len(zooms[{repr(key)})', len(val), (1, 3))
+            if len(val) == 1:
```

Comment:
Shouldn't this be before the check on the previous line that it is shape `(1, 3)`?

Reply:
The idea is that it should be one or three values, and if it's one, we promote it to three. In principle it can be done in either order, but to me this logically makes sense and follows our usual pattern: first validate that the user gave a correct input, then for our own later convenience unify the input so we know what it is (always a tuple of three float, or None).